src/Diagnostics/FxCop/Desktop.Analyzers/CSharp/MSInternal/DoNotCatchCorruptedStateExceptionsAnalyzer.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Collections.Immutable;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+using SDL.Analyzer.Common;
+
+namespace Desktop.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class CSharpDoNotCatchCorruptedStateExceptionsAnalyzer : DoNotCatchCorruptedStateExceptionsAnalyzer
+    {
+        private static readonly ImmutableArray<SyntaxKind> nodeKindsOfInterest =
+            ImmutableArray.Create(SyntaxKind.MethodDeclaration,
+                                  SyntaxKind.ConstructorDeclaration,
+                                  SyntaxKind.DestructorDeclaration,
+                                  SyntaxKind.ConversionOperatorDeclaration,
+                                  SyntaxKind.OperatorDeclaration,
+                                  SyntaxKind.AddAccessorDeclaration,
+                                  SyntaxKind.RemoveAccessorDeclaration,
+                                  SyntaxKind.GetAccessorDeclaration,
+                                  SyntaxKind.SetAccessorDeclaration);
+
+        protected override Analyzer GetAnalyzer(CompilationStartAnalysisContext context, CompilationMSInternalTypes types)
+        {
+            Analyzer analyzer = new CSharpAnalyzer(types);
+            context.RegisterSyntaxNodeAction(analyzer.AnalyzeNode, nodeKindsOfInterest);
```

Comment:
@annagrin 
If I understand the analyzer correctly, then you are basically interested in processing all catch clauses of specific exception type, and reporting diagnostics if it doesn't have any rethrow statements. Is that correct?

If so, you are better off structuring your analyzer as CodeBlock analyzer rather then a syntax node analyzer. You should do the following:

```c#
context.RegisterCodeBlockStartAction(cc =>
{
   Analyzer analyzer = new CSharpAnalyzer(types);
   cc.RegisterSyntaxNodeAction(analyzer.AnalyzeCatchClause, SyntaxKind.CatchClause);
   cc.RegisterSyntaxNodeAction(analyzer.AnalyzeThrowStatement, SyntaxKind.ThrowStatement);
   cc.RegisterCodeBlockEndAction(analyzer.OnCodeBlockEnd, SyntaxKind.ThrowStatement);
}
```

Do not use DescendantNodes to explicit traverse nodes. Instead gather your analysis state for visited catch clauses and throw statements in AnalyzeCatchClause and AnalyzeThrowStatement, and use the final state to do analysis in OnCodeBlockEnd, which will be called at the end of the method.

Here is a sample CodeBlock analyzer for your reference: https://github.com/dotnet/roslyn/blob/master/src/Samples/CSharp/Analyzers/CSharpAnalyzers/CSharpAnalyzers/StatefulAnalyzers/CodeBlockStartedAnalyzer.cs